Event open sessions handling - performance improvements (WIP)

eri-trabiccolo opened the pull request:
<!--
Contributors:
Prior to opening a pull request, please review our contributing guidelines at https://github.com/gocodebox/lifterlms/blob/trunk/.github/CONTRIBUTING.md
-->

## Description
The main purpose of this PR is to improve the perfomances when getting the open sessions to close (each 5 minutes with cron) on large websites.
The idea is to store the open sessions in a new table: `lifterlms_events_open_sessions`.
Meaning that, when opening a session we store the event, then create a new record with the session’s event `id` in this new table.
When ending a session we record this event in the events table, and remove the open session record from `lifterlms_events_open_sessions`: we always know the event id of the "starting" session:
https://github.com/gocodebox/lifterlms/blob/4.4.4/includes/class-llms-sessions.php#L119
This way the get_open_sessions() https://github.com/gocodebox/lifterlms/blob/4.4.4/includes/class-llms-sessions.php#L284 would just mean getting ids from a pretty small table (`lifterlms_events_open_sessions`).

It also fixes the few bugs I've discovered while working on this:
Fixes #1349 
Fixes #1350 
Fixes #1351
Fixes #1359 

----

This is **still WIP** as it misses the db update routine, but if you get the chance @thomasplevy I'd appreciate having your opinion on the few additions I did, if you think you would like to see additional tests on a specific area and so on.



## How has this been tested?
New and old unit tests.
But I haven't written any new test on the new open sessions handling. I relied on the old tests for the already existing methods which now leverage it.

## Screenshots <!-- if applicable -->
n/a

## Types of changes
Bug fix (non-breaking change which fixes an issue)

## Checklist:
- [ ] My code has been tested.
- [ ] My code passes all existing automated tests. <!-- Check code: `composer run-script tests-run`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/tests/README.md -->
- [ ] My code follows the LifterLMS Coding & Documentation Standards. <!-- Check code: `composer run-script check-cs-errors`, Guidelines: https://github.com/gocodebox/lifterlms/blob/trunk/docs/coding-standards.md and https://github.com/gocodebox/lifterlms/blob/trunk/docs/documentation-standards.md -->

eri-trabiccolo committed:
Introduce wp_lifterlms_events_open_sessions table where we store the open event sessions
This will highly reduce the expensivity of the query that retrieves the open event sessions,
that runs on cron every 5 minutes, to close them if idle.

eri-trabiccolo committed:
Fixed event session start not recorded on sign-on, and session end not receorded on sign-off
fixes #1349 and #1350

eri-trabiccolo committed:
Fixed event account sign-out recorded when visiting yoursite.domain/wp-admin/ as logged out user
fixes #1351

thomasplevy commented on includes/class-llms-sessions.php:
```diff
@@ -131,14 +132,28 @@ protected function end( $start ) {
 
+		if ( ! is_wp_error( $end ) ) {
+			global $wpdb;
+			$wpdb->query( // phpcs:ignore: WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
```
Just learned this cool shorthand implemented by the WPCS project to make ignoring these warnings shorter.

`WordPress.DB.DirectDatabaseQuery.DirectQuery` = `db call ok`
`WordPress.DB.DirectDatabaseQuery.NoCaching` = `no-cache ok`

```suggestion
			$wpdb->query( // db call ok; no-cache ok.
```

thomasplevy commented on includes/class-llms-sessions.php:
```diff
@@ -276,38 +299,33 @@ protected function get_last_session() {
 
 		global $wpdb;
-		$sessions = $wpdb->get_col(
+		$sessions = $wpdb->get_col( // phpcs:ignore: WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
```
```suggestion
		$sessions = $wpdb->get_col( // db call ok; no-cache ok.
```

thomasplevy commented on includes/class-llms-sessions.php:
```diff
@@ -407,26 +430,44 @@ protected function get_new_id() {
 
+		if ( ! is_wp_error( $start ) ) {
+			global $wpdb;
+			$wpdb->query( // phpcs:ignore: WordPress.DB.DirectDatabaseQuery.DirectQuery, WordPress.DB.DirectDatabaseQuery.NoCaching
```
```suggestion
			$wpdb->query( // db call ok; no-cache ok.
```

thomasplevy reviewed (commented):
Looks good so far, minor updates related to PHPCS ignores (for your information)

eri-trabiccolo committed:
Update includes/class-llms-sessions.php

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Update includes/class-llms-sessions.php

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Update includes/class-llms-sessions.php

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Added db update functions

thomasplevy commented:
@eri-trabiccolo are you ready for a review or is this still a WIP? (Not rushing, just making sure I'm not a roadblock)

Whether or not your asking for feedback, I'd like tests for the update functions. Thank you <3

eri-trabiccolo commented:
@thomasplevy 
Functionally it only misses the update function tests as you noted, and that's the reason why I hadn't asked for a review - I know they're important. <3

thomasplevy commented:
@eri-trabiccolo cool cool, no rush, just wanted to make sure I wasn't blocking!

DonavanMartin commented:
Do this pull request resolve SELECT events performance issues has showing on the following images or I need to open another issue? 
<img width="955" alt="Capture d’écran, le 2020-09-28 à 21 53 27" src="https://user-images.githubusercontent.com/7544240/94504356-95a24180-01d6-11eb-9522-377099ddc6b3.png">
<img width="917" alt="Capture d’écran, le 2020-09-28 à 21 54 20" src="https://user-images.githubusercontent.com/7544240/94504365-989d3200-01d6-11eb-9b1f-b5cf2c3e7db3.png">

thomasplevy commented:
@DonavanMartin the queries we're reworking in this PR are *more than likely* the queries responsible for the purple spikes seen in these images.

It's hard to be 100% positive as these performance graphs don't exactly report what queries are running slowly. If you want 100% confirmation a slow query log would be more useful. If you have some more concrete data for us to look at, please post it.

We don't have an open issue for this particular PR but you're welcome to open an issue if you'd like us to investigate this further. But please provide some logs in favor of graphs!

Thanks

eri-trabiccolo committed:
Add initial tests on the update functions

DonavanMartin commented:
@thomasplevy This image show the query:

<img width="886" alt="Capture d’écran, le 2020-09-29 à 16 48 57" src="https://user-images.githubusercontent.com/7544240/94621798-fb4a0880-027e-11eb-86a1-0a5abbe9ae82.png">

thomasplevy commented:
@DonavanMartin this is the query we're working on

Edit: Here's the most relevant part of the diff: https://github.com/gocodebox/lifterlms/pull/1353/files#diff-28c4ad2a7188dac0039c4e8363e4d28bL287-L301

thomasplevy commented:
@eri-trabiccolo I noticed you chose 4.4.5 for the update function prefixes, lets slate this for version 4.5.0 (barring anything crazy goes on before this is ready). I think this change warrants a minor bump.

eri-trabiccolo committed:
changes warrant a minor bump

eri-trabiccolo committed:
Added tests on update function correctly handling query "pagination"

eri-trabiccolo commented:
@thomasplevy when you have the chance can you take a look at this?

thomasplevy commented on tests/phpunit/unit-tests/functions/updates/class-llms-test-functions-updates-450.php:
```diff
@@ -0,0 +1,219 @@
+		$loops = 1;
+		// Check how many times a the update function needs to run.
+		// Internally we fetch 200 sessions at time, we expect it to run the following number of times:
+		$expected_loops = (int) ceil( count( $open_session_ids ) / 200 ) + 1;
```
Code golf:

```suggestion
		$expected_loops = 3;
```

eri-trabiccolo replied:
Ok, I just wanted to make it parametric... but ok. :)

thomasplevy reviewed (changes requested):
@eri-trabiccolo one minor (kind of joke) change requested below.

However, I think we need a bit more work here as this does not fix #1349

The tests pass, of course, but I'm not seeing the sign-on event recorded when creating a new account or logging into an existing one.

I think the creation is that we don't use WP core signon during a registration so the hook `wp_login` never fires however I should be seeing a session opened during a regular login. I'm not seeing that happen (still)

The result is that all the tests pass and the code review looks good but when I actually checkout the PR and watch changes to the DB I never see any sessions opened in the new table.

eri-trabiccolo committed:
Update tests/phpunit/unit-tests/functions/updates/class-llms-test-functions-updates-450.php

apply suggestion from code review

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo commented:
Ohh I see what you mean.
This actually fixes #1349 because it makes sure to record an open session prior to record a sign-on event, if there's no current session open. I also testes it manually - checking the db - loggin in from wp-login.php.
What you're describing here seems to be another bug then, basically autenticating via our registration and our login form doesn't not trigger the hook the wp-login hook. Right?
I'm going to look into this and see what other hook we can use, in place of, or in addition to wp-login.

eri-trabiccolo commented:
@thomasplevy 
about your latest found, I've opened a new issue https://github.com/gocodebox/lifterlms/issues/1359
but I'm going to include the fix here.

eri-trabiccolo committed:
Fix account.signon event not recorded when logging in through our login form

eri-trabiccolo committed:
Fixed account.signon event not recorded when registering through our registration/checkout forms

thomasplevy commented on includes/class.llms.person.handler.php:
```diff
@@ -21,34 +21,31 @@ class LLMS_Person_Handler {
-		/**
-		 * Allow custom username generation
-		 */
+		// Allow custom username generation.
```
```suggestion
		/*
		 * Provide a custom username generation function
		 * 
		 * If this filter returns a string, that string will be used for the student's username
		 * rather than the username automatically generated by LifterLMS.
		 *
		 * @since Unknown
		 *
		 * @param null|string $username A `null` value will cause a username to be generated, otherwise the value will be used as the username.
		 * @param string      $email    The user's email address.
		 */	
```

thomasplevy commented on includes/class.llms.person.handler.php:
```diff
@@ -759,6 +764,15 @@ public static function register( $data = array(), $screen = 'registration', $sig
 			// Signon.
 			if ( $signon ) {
 				llms_set_person_auth_cookie( $person_id, false );
+				LLMS()->events()->record(
+					array(
+						'actor_id'     => $person_id,
+						'object_type'  => 'user',
+						'object_id'    => $person_id,
+						'event_type'   => 'account',
+						'event_action' => 'signon',
+					)
+				);
```
I don't like this.

I know this fixes the issue but on reviewing the issue in #1359 it looks to me like there's actually no real reason for `llms_set_person_auth_cookie()` to exist and we could be more generally compatible with the WP core (and other plugins) by rewriting this as:

```suggestion
			// Signon.
			if ( $signon ) {
				wp_signon(
					array(
						'user_login'    => $data['user_login'],
						'user_password' => $data['password'],
					),
					is_ssl()
				);
			}
```

Since we set the `llms_last_login` usermeta data in the `llms_set_person_auth_cookie()` method we lose this data in this change.

However, in looking into this I found that we actually don't set this last login during an actual login, only during a registration.

So we could add a simple function (attached to the `wp_login` action) that will record this metadata and then we will regain this during registration and also add it during an actual login (via WP or LifterLMS or any other plugin that may take over login, say WooCommerce which we actually have an issue with here in that this metadata would never be recorded when using LifterLMS with this plugin). 

We can deprecate a function this way. What do you think?

I ran a quick test of this locally and I don't see any real issues with this change.

eri-trabiccolo replied:
@thomasplevy agreed.
I honestly tend to be conservative, even if I thought to not call `llms_set_person_auth_cookie()` too.
About the `llms_last_login` meta mmm I would just get rid of its setting. Where do we use it? Since we only set it on registration it's not even a last login, but a first login... 
Any statistic based on it would be fake, I think we use it only as fall-back in the last-seen column (which leverages the account.* events)...

thomasplevy replied:
@eri-trabiccolo 

> About the llms_last_login meta mmm I would just get rid of its setting. Where do we use it? Since we only set it on registration it's not even a last login, but a first login...
Any statistic based on it would be fake, I think we use it only as fall-back in the last-seen column (which leverages the account.* events)...

I think that at some point it wasn't fake and got lost in an unrelated fix or refactor at some point.

But yes, I think you're right we can probably drop it all together.

eri-trabiccolo replied:
@thomasplevy 
The `last_login` is used in two places:
- users table
- as fall back for "last seen" in the report's students table 

I think we want to keep this information separated from the "last seen", and that we want to avoid "calculating" it each time. Also, the `last_login` being a user meta, when populating the users table, is retrieved along with the other user metas and cache by wp in a single query.
So I think I'm going to follow your first "advice": attach the usermeta recording (with the current value `current_time('mysql')`) to the `wp_login` action, although I'm not exactly sure where to put this :D

thomasplevy commented:
@eri-trabiccolo sorry for nitpicking on this one...

eri-trabiccolo committed:
Update includes/class.llms.person.handler.php

add filter doc

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Update includes/class.llms.person.handler.php

use `wp_signon()` in favor of `llms_set_person_auth_cookie()` on registration.

Co-authored-by: Thomas Patrick Levy <thomas@gocodebox.com>

eri-trabiccolo committed:
Fix docs, pass remember=false when loggin in

eri-trabiccolo committed:
Set last login time user meta on wp_login action hook

eri-trabiccolo commented:
@thomasplevy I'm not sure that's a good place where to add the `wp_login` action.
https://github.com/gocodebox/lifterlms/pull/1353/commits/4f6dc4b6816d0fe1637991f3f6029119156c4d7c

let me know
